whetstone.py

```diff
@@ -32,22 +59,482 @@ def _authorize(self):
+            os.remove(self.filename)
+        with open(self.filename, "w") as f:
+            json.dump(data, f, indent=2)
+
+    def transform_and_load(self):
+        """Formats raw request data into relational table models and inserts into the db."""
+        data = self.get_all()
+        models = self._preprocess_records(data)
+        for model, records in models.items():
+            if records:
+                df = pd.DataFrame(records)
+                df = df.astype("object")
+                df = self._convert_dates(df)
+                df.rename(columns={"_id": "id"}, inplace=True)
+                self._write_to_db(df, model)
+
+    def _preprocess_records(self, records):
+        """
+        Any preprocessing that needs to be done to records before they are mapped to
+        models. Intended to be overridden by subclasses as needed.
+        """
+        return records
+
+    def _convert_dates(self, df):
+        """Convert date columns to the actual date time."""
+        date_types = {col: "datetime64[ns]" for col in df.columns if col in self.dates}
+        if date_types:
+            df = df.astype(date_types)
+        return df
+
+
+class Users(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "activeDistrict",
+            "archivedAt",
+            "created",
+            "email",
+            "inactive",
+            "lastActivity",
+            "lastModified",
+            "locked",
+            "name",
+            "first",
+            "last",
+            "school",
+            "course",
+            "coach",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Users": []}
+        for record in records:
+            record["school"] = record.get("defaultInformation").get("school")
+            record["course"] = record.get("defaultInformation").get("course")
+            user = {k: v for (k, v) in record.items() if k in self.columns}
+            models["Users"].append(user)
+        return models
+
+
+class Schools(Whetstone):
+    def __init__(self, sql, qa=False):
+        super().__init__(sql)
+        self.columns = [
+            "_id",
+            "internalId",
+            "name",
+            "abbreviation",
+            "archivedAt",
+            "principal",
+            "gradeSpan",
+            "lowGrade",
+            "highGrade",
+            "district",
+            "phone",
+            "address",
+            "city",
+            "cluster",
+            "region",
+            "state",
+            "zip",
+            "lastModified",
+        ]
+
+    def _preprocess_records(self, records):
+        models = {"Schools": [], "ObservationGroups": [], "ObservationGroupMembers": []}
+        for record in records:
+            record_id = record.get("_id")
+            school = {k: v for (k, v) in record.items() if k in self.columns}
```

Comment:
It feels a little strange that `self.columns` represents only one of the models. Now that there are no references to `self.columns` in the parent object, perhaps we can eliminate the subclassing of this variable and call this `self.school_columns` or something similar?

Reply:
@zkagin Every subclass still utilizes a set of columns to extract the root level nodes. We could rename them in each subclass, but that seems unnecessary. Perhaps a different variable name to identify their purpose more clearly? Something like `self.main_columns` or `self.root_columns`? Open to ideas.